diff --git a/vendor/sheredom/subprocess.h b/vendor/sheredom/subprocess.h
index 5e809023a..57fa383f3 100644
--- a/vendor/sheredom/subprocess.h
+++ b/vendor/sheredom/subprocess.h
@@ -1205,7 +1205,11 @@ cleanup:
 
   // Set working directory
   if (process_cwd) {
-#if defined(__APPLE__) && MAC_OS_X_VERSION_MIN_REQUIRED >= 260000
+#if defined(__MVS__)
+	  // posix_spawn_file_actions_addchdir_np is not available on z/OS
+	  // process_cwd is not used in llama.cpp's subprocess invocations
+	  (void)process_cwd;
+#elif defined(__APPLE__) && MAC_OS_X_VERSION_MIN_REQUIRED >= 260000
     posix_error = posix_spawn_file_actions_addchdir(&actions, process_cwd);
 #else
 #if defined(__APPLE__) && defined(__clang__)
