diff --git a/examples/gguf/gguf.cpp b/examples/gguf/gguf.cpp
index f31989c8..94ac3e5a 100644
--- a/examples/gguf/gguf.cpp
+++ b/examples/gguf/gguf.cpp
@@ -87,6 +87,7 @@ static bool gguf_ex_read_0(const std::string & fname) {
     struct gguf_init_params params = {
         /*.no_alloc = */ false,
         /*.ctx      = */ NULL,
+	/*.allow_byteswapping = */ true,
     };
 
     struct gguf_context * ctx = gguf_init_from_file(fname.c_str(), params);
