diff --git a/ggml/src/ggml-backend-reg.cpp b/ggml/src/ggml-backend-reg.cpp
index f0cdac31..29247c6a 100644
--- a/ggml/src/ggml-backend-reg.cpp
+++ b/ggml/src/ggml-backend-reg.cpp
@@ -561,7 +561,9 @@ static ggml_backend_reg_t ggml_backend_load_best(const char * name, bool silent,
 }
 
 void ggml_backend_load_all() {
-    ggml_backend_load_all_from_path(nullptr);
+#ifdef GGML_BACKEND_DL
+     ggml_backend_load_all_from_path(nullptr);
+#endif
 }
 
 void ggml_backend_load_all_from_path(const char * dir_path) {
