diff --git a/gguf-py/gguf/gguf_writer.py b/gguf-py/gguf/gguf_writer.py
index 4f23f9b0..d1fd51e0 100644
--- a/gguf-py/gguf/gguf_writer.py
+++ b/gguf-py/gguf/gguf_writer.py
@@ -4,6 +4,7 @@ import logging
 import os
 import shutil
 import struct
+import sys
 import tempfile
 from dataclasses import dataclass
 from enum import Enum, auto
@@ -452,6 +453,11 @@ class GGUFWriter:
                 for ti in tensors.values():
                     assert ti.tensor is not None  # can only iterate once over the tensors
                     assert ti.tensor.nbytes == ti.nbytes
+                    
+                    if (self.endianess == GGUFEndian.BIG and sys.byteorder == 'little') or (self.endianess == GGUFEndian.LITTLE and sys.byteorder == 'big'):
+                        # ti.tensor.byteswap(inplace=True) just didn't work here
+                        ti.tensor = ti.tensor.byteswap()
+
                     ti.tensor.tofile(fout)
                     if shard_bar is not None:
                         shard_bar.update(ti.nbytes)
