diff --git a/src/llama-context.cpp b/src/llama-context.cpp
index 06e93b19..0db8530f 100644
--- a/src/llama-context.cpp
+++ b/src/llama-context.cpp
@@ -1178,12 +1178,17 @@ int llama_context::decode(const llama_batch & batch_inp) {
                     }
                 }
             }
-
+	
+            #ifndef __MVS__
             std::fill(output_ids.begin(), output_ids.end(), -1);
-
+	    #else
+	    for (auto& id: output_ids) {
+		id = -1;
+	    }
             for (uint32_t i = 0; i < n_outputs; ++i) {
                 output_ids[out_ids[i]] = i;
             }
+	    #endif
         }
     }
 
@@ -1264,8 +1269,10 @@ uint32_t llama_context::output_reserve(int32_t n_outputs) {
     embd   = has_embd   ? output_base + logits_size : nullptr;
 
     // set all ids as invalid (negative)
-    std::fill(output_ids.begin(), output_ids.end(), -1);
-
+    //std::fill(output_ids.begin(), output_ids.end(), -1);
+    for (auto& id: output_ids) {
+	id = -1;
+    }
     this->n_outputs = 0;
 
     return n_outputs_max;
