diff --git a/ggml/src/ggml-cpu/repack.cpp b/ggml/src/ggml-cpu/repack.cpp
index 74c1c029..134b6daa 100644
--- a/ggml/src/ggml-cpu/repack.cpp
+++ b/ggml/src/ggml-cpu/repack.cpp
@@ -1424,35 +1424,37 @@ static const ggml::cpu::tensor_traits * ggml_repack_get_optimal_repack_type(cons
     // instance for IQ4
     static const ggml::cpu::repack::tensor_traits<block_iq4_nl, 4, 4, GGML_TYPE_Q8_0> iq4_nl_4x4_q8_0;
 
-    if (cur->type == GGML_TYPE_Q4_0) {
-        if (ggml_cpu_has_avx2() || (ggml_cpu_has_sve() && ggml_cpu_has_matmul_int8() && ggml_cpu_get_sve_cnt() == QK8_0)) {
-            if (cur->ne[1] % 8 == 0) {
-                return &q4_0_8x8_q8_0;
-            }
-        }
-        if (ggml_cpu_has_neon() && ggml_cpu_has_matmul_int8()) {
-            if (cur->ne[1] % 4 == 0) {
-                return &q4_0_4x8_q8_0;
-            }
-        }
-        if (ggml_cpu_has_neon() && ggml_cpu_has_dotprod()) {
-            if (cur->ne[1] % 4 == 0) {
-                return &q4_0_4x4_q8_0;
-            }
-        }
-    } else if (cur->type == GGML_TYPE_Q4_K) {
-        if (ggml_cpu_has_avx2()) {
-            if (cur->ne[1] % 8 == 0) {
-                return &q4_K_8x8_q8_K;
-            }
-        }
-    } else if (cur->type == GGML_TYPE_IQ4_NL) {
-        if (ggml_cpu_has_neon() && ggml_cpu_has_dotprod()) {
-            if (cur->ne[1] % 4 == 0) {
-                return &iq4_nl_4x4_q8_0;
-            }
-        }
-    }
+    // if (cur->type == GGML_TYPE_Q4_0) {
+    //     if (ggml_cpu_has_avx2() || (ggml_cpu_has_sve() && ggml_cpu_has_matmul_int8() && ggml_cpu_get_sve_cnt() == QK8_0)) {
+    //         if (cur->ne[1] % 8 == 0) {
+    //             return &q4_0_8x8_q8_0;
+    //         }
+    //     }
+    //     if (ggml_cpu_has_neon() && ggml_cpu_has_matmul_int8()) {
+    //         if (cur->ne[1] % 4 == 0) {
+    //             return &q4_0_4x8_q8_0;
+    //         }
+    //     }
+    //     if (ggml_cpu_has_neon() && ggml_cpu_has_dotprod()) {
+    //         if (cur->ne[1] % 4 == 0) {
+    //             return &q4_0_4x4_q8_0;
+    //         }
+    //     }
+    // } else if (cur->type == GGML_TYPE_Q4_K) {
+    //     if (ggml_cpu_has_avx2()) {
+    //         if (cur->ne[1] % 8 == 0) {
+    //             return &q4_K_8x8_q8_K;
+    //         }
+    //     }
+    // } else if (cur->type == GGML_TYPE_IQ4_NL) {
+    //     if (ggml_cpu_has_neon() && ggml_cpu_has_dotprod()) {
+    //         if (cur->ne[1] % 4 == 0) {
+    //             return &iq4_nl_4x4_q8_0;
+    //         }
+    //     }
+    // }
+
+    GGML_LOG_DEBUG("REPACK DEBUG: %s: Repacking disabled for tensor '%s' (type %s)\n",  __func__, ggml_get_name(cur) ? ggml_get_name(cur) : "unnamed", ggml_type_name(cur->type));
 
     return nullptr;
 }
