diff --git a/tests/test-gguf.cpp b/tests/test-gguf.cpp
index 3f0c312e..f7c66fb6 100644
--- a/tests/test-gguf.cpp
+++ b/tests/test-gguf.cpp
@@ -715,6 +715,7 @@ static std::pair<int, int> test_handcrafted_file(const unsigned int seed) {
         struct gguf_init_params gguf_params = {
             /*no_alloc =*/ false,
             /*ctx      =*/ hft >= offset_has_data ? &ctx : nullptr,
+	    /*.allow_byteswapping = */ true,
         };
 
         struct gguf_context * gguf_ctx = gguf_init_from_file_impl(file, gguf_params);
@@ -1111,6 +1112,7 @@ static std::pair<int, int> test_roundtrip(ggml_backend_dev_t dev, const unsigned
     struct gguf_init_params gguf_params = {
         /*no_alloc =*/ false,
         /*ctx      =*/ only_meta ? nullptr : &ctx_1,
+	/*.allow_byteswapping = */ true,
     };
     struct gguf_context * gguf_ctx_1 = gguf_init_from_file_impl(file, gguf_params);
 
