diff --git a/common/download.cpp b/common/download.cpp
index 40f6eb7..d608280 100644
--- a/common/download.cpp
+++ b/common/download.cpp
@@ -20,6 +20,9 @@
 #include <string>
 #include <thread>
 #include <vector>
+#ifdef __MVS__
+#include <cstdlib>
+#endif
 
 #include "http.h"
 
@@ -32,6 +35,8 @@
 #   endif
 #elif defined(_AIX)
 #include <sys/limits.h>
+#elif defined(__MVS__)
+#include <limits.h>
 #else
 #include <sys/syslimits.h>
 #endif
