diff --git a/common/download.cpp b/common/download.cpp
index 5ef60a4..f53ac67 100644
--- a/common/download.cpp
+++ b/common/download.cpp
@@ -18,6 +18,9 @@
 #include <string>
 #include <thread>
 #include <vector>
+#ifdef __MVS__
+#include <cstdlib>
+#endif
 
 #include "http.h"
 
@@ -30,6 +33,8 @@
 #   endif
 #elif defined(_AIX)
 #include <sys/limits.h>
+#elif defined(__MVS__)
+#include <limits.h>
 #else
 #include <sys/syslimits.h>
 #endif
