diff --git i/gguf-py/gguf/gguf_writer.py w/gguf-py/gguf/gguf_writer.py
index 010dfee..bdc1685 100644
--- i/gguf-py/gguf/gguf_writer.py
+++ w/gguf-py/gguf/gguf_writer.py
@@ -466,6 +466,11 @@ class GGUFWriter:
                 for ti in tensors.values():
                     assert ti.tensor is not None  # can only iterate once over the tensors
                     assert ti.tensor.nbytes == ti.nbytes
+                    
+                    if (self.endianess == GGUFEndian.BIG and sys.byteorder == 'little') or (self.endianess == GGUFEndian.LITTLE and sys.byteorder == 'big'):
+                        # ti.tensor.byteswap(inplace=True) just didn't work here
+                        ti.tensor = ti.tensor.byteswap()
+
                     ti.tensor.tofile(fout)
                     if shard_bar is not None:
                         shard_bar.update(ti.nbytes)
