diff --git a/tools/server/server-context.cpp b/tools/server/server-context.cpp
index bdfa517..7a61b61 100644
--- a/tools/server/server-context.cpp
+++ b/tools/server/server-context.cpp
@@ -900,6 +900,11 @@ private:
             return false;
         }
 
+        if (ctx_tgt == nullptr) {
+            SRV_ERR("failed to create context with model '%s'\n", params_base.model.path.c_str());
+            return false;
+        }
+
         vocab = llama_model_get_vocab(model_tgt);
 
         n_ctx = llama_n_ctx(ctx_tgt);
