diff --git i/vendor/cpp-httplib/httplib.h w/vendor/cpp-httplib/httplib.h
index 6ec949a..954ac6a 100644
--- i/vendor/cpp-httplib/httplib.h
+++ w/vendor/cpp-httplib/httplib.h
@@ -271,6 +271,9 @@ using socklen_t = int;
 #endif
 #ifdef __MVS__
 #include <strings.h>
+#include <zos-tls.h>
+#include <sys/time.h>
+#include <regex>
 #ifndef NI_MAXHOST
 #define NI_MAXHOST 1025
 #endif
@@ -500,6 +503,16 @@ using socket_t = int;
 /*
  * Declaration
  */
+#ifdef __MVS__
+#include <strings.h>
+#include <zos-tls.h>
+#include <sys/time.h>
+#include <regex>
+#ifndef NI_MAXHOST
+#define NI_MAXHOST 1025
+#endif
+#endif
+
 namespace httplib {
 
 namespace ws {
