diff --git i/src/llama-adapter.cpp w/src/llama-adapter.cpp
index d6a5800..1422ab4 100644
--- i/src/llama-adapter.cpp
+++ w/src/llama-adapter.cpp
@@ -153,6 +153,7 @@ static void llama_adapter_lora_init_impl(llama_model & model, const char * path_
     gguf_init_params meta_gguf_params = {
         /* .no_alloc = */ true,
         /* .ctx      = */ &ctx_init,
+	/* .allow_byteswapping = */ true,
     };
 
     gguf_context_ptr ctx_gguf { gguf_init_from_file(path_lora, meta_gguf_params) };
