diff --git i/src/llama-context.cpp w/src/llama-context.cpp
index eee9021..b95a2f4 100644
--- i/src/llama-context.cpp
+++ w/src/llama-context.cpp
@@ -1750,12 +1750,17 @@ int llama_context::decode(const llama_batch & batch_inp) {
                 // remember the swaps and apply them lazily upon logits/embeddings access
                 output_swaps.push_back({ i, j_min });
             }
-
+	
+            #ifndef __MVS__
             std::fill(output_ids.begin(), output_ids.end(), -1);
-
+	    #else
+	    for (auto& id: output_ids) {
+		id = -1;
+	    }
             for (uint32_t i = 0; i < n_outputs; ++i) {
                 output_ids[out_ids[i]] = i;
             }
+	    #endif
         }
     }
 
@@ -1890,7 +1895,9 @@ uint32_t llama_context::output_reserve(int32_t n_outputs) {
     }
 
     // set all ids as invalid (negative)
-    std::fill(output_ids.begin(), output_ids.end(), -1);
+    for (auto & id : output_ids) {
+        id = -1;
+    }
 
     this->n_outputs = 0;
 
