diff --git a/src/llama-context.cpp b/src/llama-context.cpp
index 9a40c43..8caebbf 100644
--- a/src/llama-context.cpp
+++ b/src/llama-context.cpp
@@ -3426,6 +3426,13 @@ llama_context * llama_init_from_model(
         }
     }
 
+#if defined(__MVS__)
+    if (params.flash_attn_type == LLAMA_FLASH_ATTN_TYPE_AUTO) {
+        LLAMA_LOG_INFO("%s: flash_attn auto is disabled on z/OS\n", __func__);
+        params.flash_attn_type = LLAMA_FLASH_ATTN_TYPE_DISABLED;
+    }
+#endif
+
     if (params.flash_attn_type != LLAMA_FLASH_ATTN_TYPE_DISABLED && ggml_is_quantized(params.type_k)) {
         const uint32_t blck_size = ggml_blck_size(params.type_k);
         for (uint32_t il = 0; il < model->hparams.n_layer(); ++il) {
