diff --git a/src/llama-model-loader.cpp b/src/llama-model-loader.cpp
index 510586e96..86d2e215b 100644
--- a/src/llama-model-loader.cpp
+++ b/src/llama-model-loader.cpp
@@ -551,6 +551,7 @@ llama_model_loader::llama_model_loader(
         struct gguf_init_params params = {
             /*.no_alloc = */ true,
             /*.ctx      = */ &ctx,
+	    /*.allow_byteswapping = */ true,
         };
 
         metadata_ptr.reset(gguf_init_from_file(fname.c_str(), params));
@@ -612,6 +613,7 @@ llama_model_loader::llama_model_loader(
                 struct gguf_init_params split_params = {
                     /*.no_alloc = */ true,
                     /*.ctx      = */ &ctx,
+		    /*.allow_byteswapping = */ true,
                 };
                 gguf_context_ptr ctx_gguf { gguf_init_from_file(fname_split, split_params) };
                 if (!ctx_gguf) {
@@ -663,6 +665,7 @@ llama_model_loader::llama_model_loader(
         struct gguf_init_params params = {
             /*.no_alloc = */ true,
             /*.ctx      = */ &ctx,
+	    /*.allow_byteswapping = */ true,
         };
 
         metadata_ptr.reset(gguf_init_from_file_ptr(file, params));
