diff --git i/src/llama-model-loader.cpp w/src/llama-model-loader.cpp
index 413f34c..04a17af 100644
--- i/src/llama-model-loader.cpp
+++ w/src/llama-model-loader.cpp
@@ -535,8 +535,9 @@ llama_model_loader::llama_model_loader(
         // Load the main GGUF
         struct ggml_context * ctx = NULL;
         struct gguf_init_params params = {
-            /*.no_alloc = */ true,
-            /*.ctx      = */ &ctx,
+            /*.no_alloc           = */ true,
+            /*.ctx                = */ &ctx,
+            /*.allow_byteswapping = */ true,
         };
 
         metadata_ptr.reset(gguf_init_from_file(fname.c_str(), params));
@@ -610,8 +611,9 @@ llama_model_loader::llama_model_loader(
                 const char * fname_split = splits[idx].c_str();
 
                 struct gguf_init_params split_params = {
-                    /*.no_alloc = */ true,
-                    /*.ctx      = */ &ctx,
+                    /*.no_alloc           = */ true,
+                    /*.ctx                = */ &ctx,
+                    /*.allow_byteswapping = */ true,
                 };
                 gguf_context_ptr ctx_gguf { gguf_init_from_file(fname_split, split_params) };
                 if (!ctx_gguf) {
@@ -776,6 +778,9 @@ llama_model_loader::llama_model_loader(
     if (!llama_mmap::SUPPORTED) {
         LLAMA_LOG_WARN("%s: mmap is not supported on this platform\n", __func__);
         use_mmap = false;
+    } else if (gguf_needs_byteswap(metadata)) {
+	LLAMA_LOG_WARN("%s: gguf file needs byteswapping, mmap is disabled. This may affect performance.\n", __func__);
+	use_mmap = false;
     }
 
     this->use_mmap = use_mmap;
@@ -1349,6 +1354,13 @@ void llama_model_loader::load_data_for(struct ggml_tensor * cur) const {
         const auto & file = files.at(w.idx);
         file->seek(w.offs, SEEK_SET);
         file->read_raw(cur->data, ggml_nbytes(cur));
+
+	if (gguf_needs_byteswap(metadata)) {
+	    auto byteswap = ggml_get_type_traits(cur->type)->byteswap;
+	    if (byteswap != nullptr) {
+		byteswap(cur->data, ggml_nelements(cur) / ggml_blck_size(cur->type));
+	    }
+	}
     }
 
     if (check_tensors && !ggml_validate_row_data(cur->type, cur->data, ggml_nbytes(cur))) {
@@ -1520,6 +1532,14 @@ bool llama_model_loader::load_all_data(
             if (ggml_backend_buffer_is_host(cur->buffer)) {
                 file->seek(weight->offs, SEEK_SET);
                 file->read_raw(cur->data, n_size);
+
+	        if (gguf_needs_byteswap(metadata)) {
+	            auto byteswap = ggml_get_type_traits(cur->type)->byteswap;
+	            if (byteswap != nullptr) {
+		        byteswap(cur->data, ggml_nelements(cur) / ggml_blck_size(cur->type));
+	            }
+	        }
+
                 if (check_tensors) {
                     validation_result.emplace_back(std::async(std::launch::async, [cur, n_size] {
                         return std::make_pair(cur, ggml_validate_row_data(cur->type, cur->data, n_size));
@@ -1541,8 +1561,11 @@ bool llama_model_loader::load_all_data(
                     size_t bytes_read = 0;
                     size_t data_read = 0;  // Actual tensor data copied (excluding padding)
 
+                    // for byteswapping purposes ensure that there is whole number of elements in buffer
+                    const size_t buf_size_aligned = gguf_needs_byteswap(metadata) ? buffer_size - (buffer_size % ggml_blck_size(cur->type)) : buffer_size;
+
                     while (bytes_read < read_end - read_start) {
-                        size_t read_size = std::min<size_t>(buffer_size, read_end - read_start - bytes_read);
+                        size_t read_size = std::min<size_t>(buf_size_aligned, read_end - read_start - bytes_read);
 
                         // Align the destination pointer within the pinned buffer
                         uintptr_t ptr_dest_aligned = (reinterpret_cast<uintptr_t>(host_ptrs[buffer_idx]) + alignment - 1) & ~(alignment - 1);
@@ -1568,6 +1591,13 @@ bool llama_model_loader::load_all_data(
                             data_to_copy -= (read_end - (offset + n_size));
                         }
 
+                        if (gguf_needs_byteswap(metadata)) {
+                            auto byteswap = ggml_get_type_traits(cur->type)->byteswap;
+                            if (byteswap != nullptr) {
+                                byteswap(reinterpret_cast<void *>(ptr_data), data_to_copy / ggml_blck_size(cur->type));
+                            }
+                        }
+
                         // Async upload actual data to GPU
                         ggml_backend_tensor_set_async(upload_backend, cur,
                                                       reinterpret_cast<void *>(ptr_data), data_read, data_to_copy);
@@ -1583,6 +1613,12 @@ bool llama_model_loader::load_all_data(
                     read_buf.resize(n_size);
                     file->seek(weight->offs, SEEK_SET);
                     file->read_raw(read_buf.data(), n_size);
+		    if (gguf_needs_byteswap(metadata)) {
+	    		auto byteswap = ggml_get_type_traits(cur->type)->byteswap;
+	    		if (byteswap != nullptr) {
+			    byteswap(read_buf.data(), ggml_nelements(cur) / ggml_blck_size(cur->type));
+	    		}
+		    }
                     ggml_backend_tensor_set(cur, read_buf.data(), 0, n_size);
                     if (check_tensors && !ggml_validate_row_data(cur->type, read_buf.data(), n_size)) {
                         throw std::runtime_error(format("tensor '%s' has invalid data", ggml_get_name(cur)));
